expression/builtin_regexp.go

```diff
@@ -0,0 +1,1335 @@
+			occurrence = 1
+		}
+	}
+
+	if arg_num == 5 {
+		matchType, isNull, err = re.args[4].EvalString(re.ctx, row)
+		if isNull || err != nil {
+			return "", true, err
+		}
+	}
+
+	compile, err := re.genCompile(matchType)
+	if err != nil {
+		return "", true, ErrRegexp.GenWithStackByArgs(err.Error())
+	}
+
+	reg, err := compile(pat)
+	if err != nil {
+		return "", true, ErrRegexp.GenWithStackByArgs(err.Error())
+	}
+
+	if re.isBinCollation {
+		matches := reg.FindAll(bexpr, -1)
+		length := int64(len(matches))
+		if length == 0 || occurrence > length {
+			return "", true, nil
+		}
+
+		return fmt.Sprintf("0x%s", strings.ToUpper(hex.EncodeToString(matches[occurrence-1]))), false, nil
+	}
+
+	matches := reg.FindAllString(expr, -1)
+	length := int64(len(matches))
+	if length == 0 || occurrence > length {
+		return "", true, nil
+	}
+
+	return matches[occurrence-1], false, nil
+}
+
+// we need to memorize the regexp when:
+//  1. pattern and match type are constant
+//  2. pattern is const and match type is null
+//
+// return true: need, false: needless
+func (re *regexpSubstrFuncSig) needMemorization() bool {
+	return (re.args[1].ConstItem(re.ctx.GetSessionVars().StmtCtx) && (len(re.args) <= 4 || re.args[4].ConstItem(re.ctx.GetSessionVars().StmtCtx))) && !re.isMemorizedRegexpInitialized()
+}
+
+// Call this function when at least one of the args is vector
+// REGEXP_SUBSTR(expr, pat[, pos[, occurrence[, match_type]]])
+func (re *regexpSubstrFuncSig) vecEvalString(input *chunk.Chunk, result *chunk.Column) error {
+	n := input.NumRows()
+	params := make([]*regexpParam, 0, 5)
+	defer releaseBuffers(&re.baseBuiltinFunc, params)
+
+	for i := 0; i < 2; i++ {
+		param, isConstNull, err := buildStringParam(&re.baseBuiltinFunc, i, input, false)
+		if err != nil {
+			return err
+		}
+		if isConstNull {
+			fillNullStringIntoResult(result, n)
+			return nil
+		}
+		params = append(params, param)
+	}
+
+	paramLen := len(re.args)
+
+	// Handle position parameter
+	hasPosition := (paramLen >= 3)
+	param, isConstNull, err := buildIntParam(&re.baseBuiltinFunc, 2, input, !hasPosition, 1)
+	params = append(params, param)
+
+	if err != nil {
+		return err
+	}
+	if isConstNull {
+		fillNullStringIntoResult(result, n)
+		return nil
+	}
+
+	// Handle occurrence parameter
+	hasOccur := (paramLen >= 4)
+	param, isConstNull, err = buildIntParam(&re.baseBuiltinFunc, 3, input, !hasOccur, 1)
+	params = append(params, param)
+
+	if err != nil {
+		return err
+	}
+	if isConstNull {
+		fillNullStringIntoResult(result, n)
+		return nil
+	}
+
+	// Handle match type
+	hasMatchType := (paramLen == 5)
+	param, isConstNull, err = buildStringParam(&re.baseBuiltinFunc, 4, input, !hasMatchType)
+	params = append(params, param)
+
+	if err != nil {
+		return err
+	}
+	if isConstNull {
+		fillNullStringIntoResult(result, n)
+		return nil
+	}
+
+	// Check memorization
+	if re.needMemorization() {
+		// matchType must be const or null
+		matchType := params[4].getStringVal(0)
```

Comment:
> Is it possible that row size is 0, then `params[4].getStringValue(0)` will get index out of bound error?

TiDB won't execute this function when the table is empty, however I will check this to avoid "out of bound".

Reply:
> Is it possible that row size is 0, then `params[4].getStringValue(0)` will get index out of bound error?

I will refine it.